Test Failure: DataSourceTest.testXARecovery Missing entry in database.

Test Failure: com.ibm.ws.jdbc.fat.tests.DataSourceTest.testXARecovery

Seen only on RHEL8 x86 w/ IBM JDK 8 for Postgres DB.

```
testXARecovery:junit.framework.AssertionFailedError: 2021-08-08-03:04:50:322 The response did not contain "[SUCCESS]".  Full output is:"
ERROR: Caught exception attempting to call test method testXARecovery on servlet basicfat.DataSourceTestServlet
java.lang.Exception: Missing entry in database. Results: []
	at basicfat.DataSourceTestServlet.testXARecovery(DataSourceTestServlet.java:3240)
	at componenttest.app.FATServlet.doGet(FATServlet.java:71)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:746)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:443)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1011)
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:938)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1159)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:428)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:387)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:566)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:500)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:360)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:327)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:504)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:574)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:958)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1047)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:238)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1160)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(Thread.java:822)
". expected:<0> but was:<1>
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:537)
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:495)
	at componenttest.topology.utils.HttpUtils.findStringInUrl(HttpUtils.java:472)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:444)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:414)
	at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:83)
	at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:72)
	at com.ibm.ws.jdbc.fat.tests.DataSourceTest.runTest(DataSourceTest.java:120)
	at com.ibm.ws.jdbc.fat.tests.DataSourceTest.testXARecovery(DataSourceTest.java:339)
	at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:197)
	at org.testcontainers.containers.FailureDetectingExternalResource$1.evaluate(FailureDetectingExternalResource.java:30)
	at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:318)
	at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:171)
```

There are two problems with the test. First, it tries to enlist 3 resources when PostgreSQL is configured to accept at most 2,

```
Stack Dump = org.postgresql.xa.PGXAException: Error preparing transaction. prepare xid={XidImpl: formatId(57415344), gtrid_length(36), bqual_length(54),
data(0000017c47fab2fe000000011fe97e601d006968ccde8a2675d3ca1fca5c0b69ac5a8bf00000017c47fab2fe000000011fe97e601d006968ccde8a2675d3ca1fca5c0b69ac5a8bf0000000010000000000000000000000000001)}
	at org.postgresql.xa.PGXAConnection.prepare(PGXAConnection.java:360)
	at basicfat.TestXAResource.prepare(TestXAResource.java:114)
	at com.ibm.ws.rsadapter.impl.WSRdbXaResourceImpl.prepare(WSRdbXaResourceImpl.java:695)
	at com.ibm.ejs.j2c.XATransactionWrapper.prepare(XATransactionWrapper.java:742)
	at com.ibm.tx.jta.impl.JTAXAResourceImpl.prepare(JTAXAResourceImpl.java:150)
	at com.ibm.tx.jta.impl.RegisteredResources.prepareResource(RegisteredResources.java:991)
	at com.ibm.tx.jta.impl.RegisteredResources.distributePrepare(RegisteredResources.java:1235)
	at com.ibm.tx.jta.impl.TransactionImpl.prepareResources(TransactionImpl.java:1190)
	at com.ibm.tx.jta.impl.TransactionImpl.stage1CommitProcessing(TransactionImpl.java:806)
	at com.ibm.tx.jta.impl.TransactionImpl.processCommit(TransactionImpl.java:771)
	at com.ibm.tx.jta.impl.TransactionImpl.commit(TransactionImpl.java:714)
	at com.ibm.tx.jta.impl.TranManagerImpl.commit(TranManagerImpl.java:153)
	at com.ibm.tx.jta.impl.TranManagerSet.commit(TranManagerSet.java:110)
	at com.ibm.tx.jta.impl.UserTransactionImpl.commit(UserTransactionImpl.java:145)
	at com.ibm.tx.jta.embeddable.impl.EmbeddableUserTransactionImpl.commit(EmbeddableUserTransactionImpl.java:101)
	at com.ibm.ws.transaction.services.UserTransactionService.commit(UserTransactionService.java:72)
	at basicfat.DataSourceTestServlet.testXARecovery(DataSourceTestServlet.java:3169)
...
Caused by: org.postgresql.util.PSQLException: ERROR: maximum number of prepared transactions reached
  Hint: Increase max_prepared_transactions (currently 2).
```

Secondly, it only waits up to 5 seconds for the recovery to occur.

--- a/dev/com.ibm.ws.jdbc_fat/test-applications/basicfat/src/basicfat/DataSourceTestServlet.java
+++ b/dev/com.ibm.ws.jdbc_fat/test-applications/basicfat/src/basicfat/DataSourceTestServlet.java
@@ -3124,6 +3124,7 @@ public class DataSourceTestServlet extends FATServlet {
      * The recoveryAuthData should be used for recovery.
      */
     public void testXARecovery() throws Throwable {
+        int numPrepares = 3;
         clearTable(ds4u_2);
         Connection[] cons = new Connection[3];
         tran.begin();
@@ -3135,6 +3136,12 @@ public class DataSourceTestServlet extends FATServlet {
 
             String dbProductName = cons[0].getMetaData().getDatabaseProductName().toUpperCase();
             System.out.println("Product Name is " + dbProductName);
+
+            // Work around PostgreSQL's limit of max_prepared_transactions=2,
+            // which results in the following message when exceeded:
+            // "maximum number of prepared transactions reached"
+            if ("POSTGRESQL".equalsIgnoreCase(dbProductName))
+                numPrepares = 2;
 
             // Verify isolation-level="TRANSACTION_READ_COMMITTED" from ibm-web-ext.xml
             int isolation = cons[0].getTransactionIsolation();
@@ -3156,12 +3163,14 @@ public class DataSourceTestServlet extends FATServlet {
             pstmt.executeUpdate();
             pstmt.close();
 
-            pstmt = cons[2].prepareStatement("insert into cities values (?, ?, ?)");
-            pstmt.setString(1, "Moorhead");
-            pstmt.setInt(2, 38065);
-            pstmt.setString(3, "Clay");
-            pstmt.executeUpdate();
-            pstmt.close();
+            if (numPrepares == 3) {
+                pstmt = cons[2].prepareStatement("insert into cities values (?, ?, ?)");
+                pstmt.setString(1, "Moorhead");
+                pstmt.setInt(2, 38065);
+                pstmt.setString(3, "Clay");
+                pstmt.executeUpdate();
+                pstmt.close();
+            }
 
             System.out.println("Intentionally causing in-doubt transaction");
             TestXAResource.assignSuccessLimit(1, cons);
@@ -3206,12 +3215,12 @@ public class DataSourceTestServlet extends FATServlet {
             PreparedStatement pstmt = con.prepareStatement("select name, population, county from cities where name = ?");
 
             /*
-             * Poll for results once a second for 5 seconds.
+             * Poll for results once a second for up to 2 minutes.
              * Most databases will have XA recovery done by this point
              *
              */
             List<String> cities = new ArrayList<>();
-            for (int count = 0; cities.size() < 3 && count < 5; Thread.sleep(1000)) {
+            for (int count = 0; cities.size() < numPrepares && count < 120; Thread.sleep(1000)) {
                 if (!cities.contains("Edina")) {
                     pstmt.setString(1, "Edina");
                     result = pstmt.executeQuery();
@@ -3226,7 +3235,7 @@ public class DataSourceTestServlet extends FATServlet {
                         cities.add(1, "St. Louis Park");
                 }
 
-                if (!cities.contains("Moorhead")) {
+                if (numPrepares == 3 && !cities.contains("Moorhead")) {
                     pstmt.setString(1, "Moorhead");
                     result = pstmt.executeQuery();
                     if (result.next())
@@ -3236,7 +3245,7 @@ public class DataSourceTestServlet extends FATServlet {
                 System.out.println("Attempt " + count + " to retrieve recovered XA data. Current status: " + cities);
             }
 
-            if (cities.size() < 3)
+            if (cities.size() < numPrepares)
                 throw new Exception("Missing entry in database. Results: " + cities);
             else
                 System.out.println("successfully accessed the data");

